fms/triton/pytorch_ops.py

```diff
@@ -0,0 +1,224 @@
+)
+
+
+# All that's needed for torch.compile support
+@torch.library.impl(lib, "moe_mm", "Meta")
+def moe_mm_meta(
+    input: torch.Tensor,
+    moe_matrix: torch.Tensor,
+    token_expert_mapping: torch.Tensor,
+    padded_token_ids_per_block: torch.Tensor,
+    expert_block_mapping: torch.Tensor,
+    total_padded_tokens: torch.Tensor,
+    topk: int,
+    padding_size,
+):
+    M, A = token_expert_mapping.shape
+    _, N, _ = moe_matrix.shape
+    return torch.empty((M, A, N), device=input.device, dtype=input.dtype)
+
+
+@torch.library.impl(lib, "moe_mm", "CUDA")
+def moe_mm(
+    input: torch.Tensor,
+    moe_matrix: torch.Tensor,
+    token_expert_mapping: torch.Tensor,
+    padded_token_ids_per_block: torch.Tensor,
+    expert_block_mapping: torch.Tensor,
+    total_padded_tokens: torch.Tensor,
+    topk: int,
+    padding_size,
+):
+    from fms.triton.moe_kernel import (
+        BEST_MOE_CONFIGS,
+        _autotune,
+        _create_best_configs_key,
+        _load_best_configs,
+        _save_best_configs,
+        invoke_fused_moe_kernel,
+    )
+
+    M, A = token_expert_mapping.shape
+    E, N, _ = moe_matrix.shape
+    output = torch.zeros((M, A, N), device=input.device, dtype=input.dtype)
+
+    if BEST_MOE_CONFIGS is None:
+        BEST_MOE_CONFIGS = _load_best_configs()
+    # Loading must have not been successful. Let's create a new dictionary.
+    if BEST_MOE_CONFIGS is None:
+        BEST_MOE_CONFIGS = {}
+    key = _create_best_configs_key(input, moe_matrix, token_expert_mapping)
+    if key not in BEST_MOE_CONFIGS:
+        import functools
+
+        # TODO: Add more configs?
+        configs = [
+            {
+                "block_m": 64,
+                "block_n": 64,
+                "block_k": 32,
+            },
+            {
+                "block_m": 16,
+                "block_n": 32,
+                "block_k": 64,
+            },
+        ]
+
+        configs = [config for config in configs if config["block_m"] == padding_size]
+        best, best_config = _autotune(
+            configs,
+            functools.partial(
+                invoke_fused_moe_kernel,
+                input,
+                moe_matrix,
+                output,
+                token_expert_mapping,
+                padded_token_ids_per_block,
+                expert_block_mapping,
+                total_padded_tokens,
+                topk,
+            ),
+        )
+        BEST_MOE_CONFIGS[key] = best_config
+        _save_best_configs(BEST_MOE_CONFIGS)
+    best_config = BEST_MOE_CONFIGS[key]
+    if best_config is None:
+        return torch.tensor([])
+
+    invoke_fused_moe_kernel(
+        input,
+        moe_matrix,
+        output,
+        token_expert_mapping,
+        padded_token_ids_per_block,
+        expert_block_mapping,
+        total_padded_tokens,
+        topk,
+        best_config,
+    )
+
+    return output
+
+
+@torch.library.impl(lib, "moe_mm", "CPU")
```

Comment:
also not relevant to the particular fix here, but I wonder if we might also want to register this for mps?

Reply:
ok, feel free to test if it works as is when it's merged